### PR TITLE
Clean up orphaned instances and security groups (HMS-3632)

### PR DESCRIPTION
**Pull request by schuellerf**
This pull request includes:

- [x] adequate testing for the new functionality or fixed issue
- [x] adequate documentation informing people about the change such as
  - [ ] submit a PR for the READMEs [listed here](https://github.com/osbuild/osbuild.github.io/blob/main/readme-list)
  - [ ] submit a PR for the [osbuild.org website](https://github.com/osbuild/osbuild.github.io/) repository if this PR changed any behavior not covered by the automatically updated READMEs

Implement removal of orphaned items as described in https://issues.redhat.com/browse/HMS-3632
removal of launch templates is not yet implemented (that's why it's a draft for now)

### CODE REVIEW

File: cmd/osbuild-service-maintenance/aws.go [hunk 1]
```diff
@@ -84,13 +99,58 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
...
+	if errSecureInstances != nil {
+		logrus.Errorf("Error in terminating secure instances: %v, continuing other cleanup.", errSecureInstances)
+		err = errSecureInstances
+	}
+
+	errSecurityGroups := searchSGAndCleanup(ctx, a, dryRun)
+	if errSecurityGroups != nil {
+		logrus.Errorf("Error in cleaning up security groups: %v", errSecurityGroups)
+		if err != nil {
+			err = fmt.Errorf("Multiple errors while processing AWSCleanup: %w and %w.", err, errSecurityGroups)
```
Comment (schuellerf) [inline cmd/osbuild-service-maintenance/aws.go:116]: Is this a valid pattern, to try to continue cleanup and collect the errors?

File: cmd/osbuild-service-maintenance/aws.go [hunk 2]
```diff
@@ -103,14 +163,164 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
...
+	}
+	if len(parent) != 1 {
+		logrus.Errorf("Instance %s has %d parents. That should never happen, not changing anything here.", childId, len(parent))
+		return true
+	}
+	if len(parent[0].Instances) == 0 {
+		logrus.Infof("Instance %s has no parent instance, removing it", childId)
+		return false
+	}
+	if len(parent[0].Instances) != 1 {
```
Comment (schuellerf) [inline cmd/osbuild-service-maintenance/aws.go:220]: not sure if all those checks are necessary?

File: internal/cloud/awscloud/maintenance.go [hunk 3]
```diff
@@ -84,6 +85,31 @@ func (a *AWS) DescribeInstancesByTag(tagKey, tagValue string) ([]ec2types.Reserv
...
+func (a *AWS) DescribeInstancesByTag(tagKey, tagValue string) ([]ec2types.Reservation, error) {
+	return a.describeInstancesByKeyValue(fmt.Sprintf("tag:%s", tagKey), tagValue)
+}
+
+func (a *AWS) DescribeInstancesBySecurityGroupID(securityGroupID string) ([]ec2types.Reservation, error) {
+	return a.describeInstancesByKeyValue("instance.group-id", securityGroupID)
+}
+
+func (a *AWS) DescribeInstancesByLaunchTemplateID(launchTemplateID string) ([]ec2types.Reservation, error) {
+	return a.describeInstancesByKeyValue("tag:aws:ec2launchtemplate:id", launchTemplateID)
```
Comment (schuellerf) [inline internal/cloud/awscloud/maintenance.go:97]: This looks weird - I'll check if there is another option to get the instances (or do you know by heart how it works @croissanne ?)

Comment (croissanne) [reply internal/cloud/awscloud/maintenance.go:97]: Eh, looks ok to me. https://awscli.amazonaws.com/v2/documentation/api/2.0.33/reference/ec2/describe-instances.html#options are the filters, and i don't see launchtemplate

File: cmd/osbuild-service-maintenance/aws.go (hunk 2 above)
Comment (croissanne) [reply cmd/osbuild-service-maintenance/aws.go:222]: I tihnk it's ok

File: cmd/osbuild-service-maintenance/aws.go [hunk 4]
```diff
@@ -84,33 +99,230 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
...
+				inUse = true
+				break
+			}
+		}
+	}
+	return inUse
+}
+
+func searchLTAndCleanup(ctx context.Context, a *awscloud.AWS, dryRun bool) error {
+
```
Comment (croissanne) [inline cmd/osbuild-service-maintenance/aws.go:291]: rm newline here

**Review by croissanne — COMMENTED** [review]
In general we don't put the jira ticket in the commit message itself, just the PR title.

File: cmd/osbuild-service-maintenance/aws.go [hunk 5]
```diff
@@ -112,17 +128,73 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
...
+				if !slices.Contains(instanceIDs, *i.InstanceId) {
+					instanceIDs = append(instanceIDs, *i.InstanceId)
+				}
 			}
 		}
 	}
 	return instanceIDs
 }
+
+func getChildAndParentInstances(reservations []ec2types.Reservation) []struct {
```
Comment (croissanne) [inline cmd/osbuild-service-maintenance/aws.go:182]: I think defining this struct up to instead of an anonymous struct would make it more readable.

File: cmd/osbuild-service-maintenance/aws.go [hunk 6]
```diff
@@ -97,6 +99,28 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
...
+	// continue execution if one cleanup fails
+	err = nil
+	errSecureInstances := terminateOrphanedSecureInstances(a, dryRun)
+	// keep going with other cleanup even on error
+	if errSecureInstances != nil {
+		logrus.Errorf("Error in terminating secure instances: %v, continuing other cleanup.", errSecureInstances)
+		err = errSecureInstances
+	}
+
+	errSecurityGroups := searchSGAndCleanup(ctx, a, dryRun)
```
Comment (croissanne) [inline cmd/osbuild-service-maintenance/aws.go:112]: I'm not sure if we should share the same context here? It probably doesn't matter that much but I don't get the rationale for doing so here.

File: cmd/osbuild-service-maintenance/aws.go (hunk 1 above)
Comment (croissanne) [reply cmd/osbuild-service-maintenance/aws.go:116]: I think generally error wrapping is used to wrap deeper errors, not really make a list of them. So maybe we should just add the errors to a slice and then join them together in the end? I think it might make it more readable.

File: cmd/osbuild-service-maintenance/aws.go (hunk 4 above)
Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go:291]: Ups - thanks!

File: cmd/osbuild-service-maintenance/aws.go [hunk 7]
```diff
@@ -192,9 +215,69 @@ func checkValidParent(childId string, parent []ec2types.Reservation) bool {
...
+		} else {
+			logrus.Debugf("Security group %s has non terminated instances associated with it.", *sg.GroupId)
+		}
+	}
+	return nil
+}
+
+// checkIfInUse returns true if any instance of the reservations is not terminated
+// then it's considered "in use"
+func checkIfInUse(reservations []ec2types.Reservation) bool {
```
Comment (croissanne) [inline cmd/osbuild-service-maintenance/aws.go:276]: can we just call this "isInstanceRunning"? I thinkit makes it a bit clearer.

**Review by croissanne — CHANGES_REQUESTED** [review]
thanks, looks good overall

Could you reword this commit message though?
```
Usually the parent does not exist when also the secure instance is
older than 2h, but this an extra safeguard that secure instances
without a valid parent instance does not make sense to keep.
```

And then the jira tickets tend to just go in the PR descriptions.

File: cmd/osbuild-service-maintenance/aws.go (hunk 6 above)
Comment (croissanne) [reply cmd/osbuild-service-maintenance/aws.go:112]: Like the context isn't passed from up above, but created in this function.

**Comment by schuellerf** [comment]
> And then the jira tickets tend to just go in the PR descriptions.

I think they should be part of the commit message
https://osbuild.org/docs/developer-guide/general/workflow
but I'll move it to the "git commit subject" that's more readable for sure

File: cmd/osbuild-service-maintenance/aws.go (hunk 6 above)
Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go:112]: I didn't want to change all of the functions from above in this PR, but prepare as much as possible.
As far as I understood that's the idea of the context, that it should always be traversed from top to bottom to be able to hand over cancel, timeouts and other properties…

File: cmd/osbuild-service-maintenance/aws.go (hunk 7 above)
Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go:276]: It's more than running. Also pending, shutting-down.… but you are right, it sounds odd…

Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go:276]: or invert and `allTerminated`

File: cmd/osbuild-service-maintenance/aws.go (hunk 1 above)
Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go]: please check

File: cmd/osbuild-service-maintenance/aws.go (hunk 7 above)
Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go]: please check

File: cmd/osbuild-service-maintenance/aws.go [hunk 8]
```diff
@@ -84,33 +126,212 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
 		wg.Wait()
 	}
 
+	// using `errors` to collect all errors as we want to
+	// continue execution if only one cleanup fails
+	var errors MultiError
```
Comment (mvo5) [inline cmd/osbuild-service-maintenance/aws.go:131]: Quick drive-by, there is `errors.Join()` now, the more common way to write this (IME) is:
```go
var errs []error
...
errors = append(errs, err)
...
return errors.Join(errs)
```

File: cmd/osbuild-service-maintenance/aws.go [hunk 9]
```diff
@@ -13,10 +15,50 @@ import (
...
 )
 
-func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey string, cutoff time.Time) error {
-	a, err := awscloud.New("us-east-1", accessKeyID, accessKey, "")
+type ChildToParentAssociation struct {
+	Child  string
+	Parent string
+}
+
+type MultiError struct {
```
Comment (mvo5) [inline cmd/osbuild-service-maintenance/aws.go:23]: I think we want to use `errors.Join()` instead which gives us the stdlib "joinErrors" struct which does the same expect it also implements Unwrap (so that errors.{Is,As} keeps working)

File: cmd/osbuild-service-maintenance/aws.go [hunk 10]
```diff
@@ -84,33 +126,212 @@ func AWSCleanup(maxConcurrentRequests int, dryRun bool, accessKeyID, accessKey s
...
+
+// allTerminated returns true if any instance of the reservations is not terminated
+// then it's considered "in use"
+func allTerminated(reservations []ec2types.Reservation) bool {
+	allAreTerminated := true
+
+	for _, reservation := range reservations {
+		for _, instance := range reservation.Instances {
+			if instance.State != nil && (instance.State.Name != ec2types.InstanceStateNameTerminated) {
+				allAreTerminated = false
```
Comment (mvo5) [inline cmd/osbuild-service-maintenance/aws.go:294]: This could just `return false` here, this would avoid the need for the `allAreTerminated` var (and the last return of the function would just be `return true`)

**Review by mvo5 — COMMENTED** [review]
Thanks for working on this - quick drive-by comments/suggestions (purely on the tactical level :)

File: cmd/osbuild-service-maintenance/aws.go (hunk 6 above)
Comment (croissanne) [reply cmd/osbuild-service-maintenance/aws.go]: ack

**Review by croissanne — COMMENTED** [review]
other than  @mvo5's comments lgtm :)

File: cmd/osbuild-service-maintenance/aws.go (hunk 10 above)
Comment (schuellerf) [reply cmd/osbuild-service-maintenance/aws.go:294]: Thanks! …refactoring mess I left there :-/

**Review by croissanne — APPROVED** [review]
LGTM, ty!
